src/softmax.jl

```diff
@@ -61,23 +75,26 @@ It is semantically equivalent to the following:
 See also [`softmax`](@ref).
 """
 logsoftmax(x; dims = 1) = logsoftmax!(similar(x, (float ∘ eltype)(x)), x; dims = dims)
+
 logsoftmax!(x; dims = 1) = logsoftmax!(x, x; dims = dims)
+
 function logsoftmax!(out::O, x::T; dims = 1) where {O<:AbstractArray,T<:AbstractArray}
     out .= x .- maximum(x; dims = dims)
-    # out .-= log.(sum(exp.(out); dims = dims))  # WARN: this will decrease performance.
-    log_ = log.(sum(exp.(out); dims = dims))
-    out .-= log_
+    red = sum(exp, out; dims = dims)
+    red .= log.(red)
+    out .-= red
```

Comment:
The flaw in my bright idea here is that if `dims=:` then `sum(exp, rand(3), dims=:) isa Number`, so writing the `log.(red)` back into it won't work. (Although `dims=1:N` would be OK.) Sorry!

What's the least inelegant way to fix this? One way is
```suggestion
    sum_ = sum(exp, out; dims = dims)
    log_ = ndims(sum_) > 0 ? red .= log.(sum_) : log.(sum_)
    out .-= log_
```
and another is just to give up on recycling that array, and just say `log_  = log.(sum_)`.

Reply:
yeah maybe better not trying to be too smart here, also the  logsum output has reduced size, so probably no big deal

Reply:
Indeed, there are a bunch of O(N^2) allocations (for matrix input) so perhaps one O(N) doesn't matter at all.